bindings/rust/s2n-tls-tokio/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use errno::{set_errno, Errno};
+use s2n_tls::raw::{
+    config::Config,
+    connection::Connection,
+    error::Error,
+    ffi::{s2n_mode, s2n_status_code},
+};
+use std::{
+    future::Future,
+    os::raw::{c_int, c_void},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+pub struct TlsAcceptor {
+    config: Config,
+}
+
+impl TlsAcceptor {
+    pub fn new(config: Config) -> Self {
+        TlsAcceptor { config }
+    }
+
+    pub async fn accept<S>(&self, stream: S) -> Result<TlsStream<S>, Error>
+    where
+        S: AsyncRead + AsyncWrite + Unpin,
+    {
+        TlsStream::open(self.config.clone(), s2n_mode::SERVER, stream).await
+    }
+}
+
+pub struct TlsConnector {
+    config: Config,
+}
+
+impl TlsConnector {
+    pub fn new(config: Config) -> Self {
+        TlsConnector { config }
+    }
+
+    pub async fn connect<S>(&self, _domain: &str, stream: S) -> Result<TlsStream<S>, Error>
+    where
+        S: AsyncRead + AsyncWrite + Unpin,
+    {
+        TlsStream::open(self.config.clone(), s2n_mode::CLIENT, stream).await
+    }
+}
+
+struct TlsHandshake<'a, S> {
+    tls: &'a mut TlsStream<S>,
+}
+
+impl<S> Future for TlsHandshake<'_, S>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = Result<(), Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.tls.with_io(|context| {
+            context.conn.set_waker(Some(cx.waker()))?;
+            context.conn.negotiate().map(|r| r.map(|_| ()))
+        })
+    }
+}
+
+pub struct TlsStream<S> {
+    conn: Connection,
+    stream: S,
+}
+
+impl<S> TlsStream<S>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    async fn open(config: Config, mode: s2n_mode::Type, stream: S) -> Result<Self, Error> {
+        let mut conn = Connection::new(mode);
+        conn.set_config(config)?;
+
+        let mut tls = TlsStream { conn, stream };
+        TlsHandshake { tls: &mut tls }.await?;
+        Ok(tls)
+    }
+
+    fn with_io<F>(&mut self, action: F) -> Poll<Result<(), Error>>
+    where
+        F: FnOnce(&mut Self) -> Poll<Result<(), Error>>,
+    {
+        unsafe {
```

Comment:
can you add a safety comment here giving some context on the safety of this code?